feat(auth): Add auth emulator support via the FIREBASE_AUTH_EMULATOR_HOST environment variable.

muru opened the pull request:
### Discussion

Closes #503, by adding support for the authentication emulator via the `FIREBASE_AUTH_EMULATOR_HOST` environment variable.

Additionally, I added a change that *shouldn't* break anything, but does make the SDK work better with the emulator,  fixing the equivalent of https://github.com/firebase/firebase-admin-node/issues/1084. I ran into this while testing these changes in a project, and I feel it should be part of "supporting" the emulator, but I am open to putting it in a separate PR if it doesn't belong.

### Testing

- [x] Existing tests pass
- [x] Modified the existing tests for user management, token generation and providers to run for both emulated and non-emulated cases.

### API Changes

None, except insofar that the `FIREBASE_AUTH_EMULATOR_HOST` environment variable becomes part of the API.

yuchenshi reviewed (commented):
The approach looks right to me, although we should also handle *verifying* ID tokens and session cookies.

yuchenshi commented:
cc @brianrodri who may be interested in this.

muru commented:
@yuchenshi  At the time I started work on this I thought the Google people were going to rework how those are handled. Now that the firebase CLI is supposed to work OK with session cookies, I'll try to include it in integration tests (which I'd held off on) and see what else needs to be fixed here.

yuchenshi commented:
> @yuchenshi At the time I started work on this I thought the Google people were going to rework how those are handled. Now that the firebase CLI is supposed to work OK with session cookies, I'll try to include it in integration tests (which I'd held off on) and see what else needs to be fixed here.

I really appreciate it. https://github.com/firebase/firebase-admin-node/pull/1148 has some integration tests that you can draw inspiration from and it shows the desired behavior for verifying ID tokens.

hiranya911 commented on firebase_admin/_token_gen.py:
```diff
@@ -198,7 +200,7 @@ def create_session_cookie(self, id_token, expires_in):
         payload = {
             'idToken': id_token,
-            'validDuration': expires_in,
+            'validDuration': str(expires_in),
```
I think this is no longer necessary? @yuchenshi should be able to confirm.

yuchenshi replied:
This should be no longer necessary with the latest Firebase CLI. Please let us know if otherwise. Integration tests can help here.

muru replied:
I'll dropped this commit for now, and add it back if the integration tests have a problem (which I don't think will happen).

hiranya911 commented on tests/testutils.py:
```diff
@@ -58,6 +60,15 @@ def run_without_project_id(func):
 
+def new_monkeypatch():
+    try:
+        monkeypatch = pytest.MonkeyPatch()
```
Local variable not needed. Can just return here and below.

hiranya911 reviewed (changes requested):
Thanks for the PR @muru. Looks mostly good. I've pointed out a few things that can be improved.

Keramblock commented:
@muru sorry for disturbance, but could you please take a look? I really wait for that MR  =)

muru commented on firebase_admin/_auth_client.py:
```diff
@@ -36,17 +39,42 @@ def __init__(self, app, tenant_id=None):
+                    'Invalid {0}: "{1}". It must follow format "host:port".'.format(
+                        _EMULATOR_HOST_ENV_VAR, emulator_host))
+            # Used instead of credential
+            http_headers["Authorization"] = "Bearer owner"
```
This turned out to be a bit tricky - one of tests (`tests/test_user_mgt.py::TestImportUsers::test_import_users_http_error`) would fail trying to refresh the token. So I have settled in favour of using credentials if provided, or falling back to setting the token if not.

muru commented:
@Keramblock thanks for the nudge. I have fixed the issues noted by @hiranya911 , but haven't gotten the integration tests to work. I'll try those again over the weekend.

hiranya911 commented on firebase_admin/_token_gen.py:
```diff
@@ -84,11 +84,13 @@ class TokenGenerator:
     ID_TOOLKIT_URL = 'https://identitytoolkit.googleapis.com/v1'
 
-    def __init__(self, app, http_client):
+    def __init__(self, app, http_client, id_toolkit_url=None):
```
A little readability nit: Call the new method argument `url_override`.

```py
def __init__(self, app, http_client, url_override=None):
    url_prefix = url_override if url_override else self.ID_TOOLKIT_URL
    self.base_url = '{0}/projects/{1}'.format(url_prefix, app.project_id)
```

hiranya911 commented on firebase_admin/_auth_client.py:
```diff
@@ -36,17 +42,45 @@ def __init__(self, app, tenant_id=None):
 
         self._tenant_id = tenant_id
-        self._token_generator = _token_gen.TokenGenerator(app, http_client)
+        self._token_generator = _token_gen.TokenGenerator(
```
I'd recommend using keyword args for the last argument for clarity. But not a dealbreaker.

hiranya911 reviewed (changes requested):
I think the way credentials are being used is still not quite correct. But overall looks pretty good.

muru commented on firebase_admin/_auth_client.py:
```diff
@@ -36,17 +42,45 @@ def __init__(self, app, tenant_id=None):
+            base_url = 'http://{0}/identitytoolkit.googleapis.com'.format(emulator_host)
+            id_toolkit_endpoints['v1'] = base_url + '/v1'
+            id_toolkit_endpoints['v2beta1'] = base_url + '/v2beta1'
+
```
@hiranya911  That was how I approached it at first, but then `TestImportUsers::test_import_users_http_error` failed (well, I guess it expects to fail, but only in a particular way). I am not entirely sure of the implications of adding another expected failure mode there - that could cause real problems to creep past unnoticed.

hiranya911 replied:
What's the exact test failure you're setting? Can you share some details or a stacktrace?

hiranya911 replied:
Ok, having looked at the source I think I know what the problem is. That test case is simulating a 401 error, which forces the credential to be refreshed, and that's resulting in an error. You need to use something like the following:

https://github.com/firebase/firebase-admin-python/blob/c10c7471eda060ef16a11e568e3e503316e67793/firebase_admin/db.py#L972-L978

I'd recommend moving that class into `_utils.py`, and reusing it for both auth and db.

hiranya911 commented:
@muru the last set of commits have broken the CI. I'd recommend reverting the breaking commits, and doing the bare minimal necessary to address the earlier code review feedback. Any other new capabilities (e.g. emulator based testing), can be implemented in separate future PRs.

muru commented:
@hiranya911 porting the token verification code from https://github.com/firebase/firebase-admin-node/pull/1148 breaks a few unit tests. Since that should probably be included with this PR, would it be fine if I marked those tests for skipping when tested with the emulator? And while I am it, I could do the same for integration tests too.

hiranya911 commented:
Ideally we shouldn't have to skip any unit tests. 

We can live without emulator-based integration tests for now (that work is on hold in Node.js at https://github.com/firebase/firebase-admin-node/pull/1155 due to some problems).

muru commented:
@hiranya911 here are the tests that fail with the emulator (mostly those concerned with verifying signatures, invalid tokens, expired cookies, and the OIDC/SAML providers):

<details><summary><code>tests/test_token_gen.py</code> (39 failures)</summary>

```
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_valid_params[auth_app1-Basic] - ValueError: Could not verify token signature.
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_valid_params[auth_app1-NoDevClaims] - ValueError: Could not verify token signature.
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_valid_params[auth_app1-EmptyDevClaims] - ValueError: Could not verify token signature.
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_noncert_credential[user_mgt_app0] - Failed: DID NOT RAISE <class 'ValueError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_revoked_token_check_revoked[user_mgt_app0-BinaryToken] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.RevokedIdToke...
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_revoked_token_check_revoked[user_mgt_app0-TextToken] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.RevokedIdTokenE...
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app0-NoKid] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app0-WrongKid] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app0-FutureToken] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app0-ExpiredToken] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_expired_token[user_mgt_app0] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.ExpiredIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_certificate_request_failure[user_mgt_app0] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.CertificateFetchError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app0-NoKid] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCookieError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app0-WrongKid] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCookieErr...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app0-FutureCookie] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCooki...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app0-ExpiredCookie] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCook...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_expired_cookie[user_mgt_app0] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.ExpiredSessionCookieError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_certificate_request_failure[user_mgt_app0] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.CertificateFetchErr...
FAILED tests/test_token_gen.py::TestCertificateCaching::test_certificate_caching[user_mgt_app0] - AssertionError: assert 0 == 1
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_noncert_credential[user_mgt_app1] - Failed: DID NOT RAISE <class 'ValueError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_revoked_token_check_revoked[user_mgt_app1-BinaryToken] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.RevokedIdToke...
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_revoked_token_check_revoked[user_mgt_app1-TextToken] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.RevokedIdTokenE...
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-NoKid] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-WrongKid] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-FutureToken] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-ExpiredToken] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_expired_token[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.ExpiredIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_certificate_request_failure[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.CertificateFetchError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-NoKid] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCookieError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-WrongKid] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCookieErr...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-FutureCookie] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCooki...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-ExpiredCookie] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCook...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_expired_cookie[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.ExpiredSessionCookieError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_certificate_request_failure[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.CertificateFetchErr...
FAILED tests/test_token_gen.py::TestCertificateCaching::test_certificate_caching[user_mgt_app1] - AssertionError: assert 0 == 1
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_sign_with_iam - AssertionError: assert False
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_sign_with_iam_error - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.TokenSignError'>
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_sign_with_discovered_service_account - AssertionError: assert False
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_sign_with_discovery_failure - Failed: DID NOT RAISE <class 'ValueError'>
```
</details>

<details><summary><code>integration/test_auth.py</code> (13 failures)</summary>

```
ERROR integration/test_auth.py::test_create_oidc_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
ERROR integration/test_auth.py::test_get_oidc_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
ERROR integration/test_auth.py::test_list_oidc_provider_configs - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
ERROR integration/test_auth.py::test_create_saml_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
ERROR integration/test_auth.py::test_get_saml_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
ERROR integration/test_auth.py::test_list_saml_provider_configs - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
FAILED integration/test_auth.py::test_custom_token_without_service_account - google.auth.exceptions.RefreshError: ('invalid_grant: Invalid grant: account not found', '{"error":"inval...
FAILED integration/test_auth.py::test_verify_id_token_revoked - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.RevokedIdTokenError'>
FAILED integration/test_auth.py::test_import_users_with_password - requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: http://localhost:9099/www.googleapis.com/iden...
FAILED integration/test_auth.py::test_update_oidc_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
FAILED integration/test_auth.py::test_delete_oidc_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
FAILED integration/test_auth.py::test_update_saml_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
FAILED integration/test_auth.py::test_delete_saml_provider_config - firebase_admin.exceptions.NotFoundError: Error while calling Auth service (Not Found).
```
</details>
I have split up the recent commit, putting the changes skipping these tests (note - skipped only when the emulator is used) as a separate commit. Please have a look when you have the time - if there are some tests which mustn't be skipped with the emulator, we can fix them now. I think I have also addressed the other review comments.

hiranya911 commented:
Any idea how this happened?

```
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_noncert_credential[user_mgt_app0] - Failed: DID NOT RAISE <class 'ValueError'>
...
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_noncert_credential[user_mgt_app1] - Failed: DID NOT RAISE <class 'ValueError'>
```

Shouldn't one of them pass?

hiranya911 reviewed (changes requested):
Let's not implement integration testing for emulator mode yet. Lets focus on getting the implementation and the unit tests absolutely correct.

muru commented:
@hiranya911 what you say is correct, but shouldn't matter here, since emulation is set in the second/final iteration (https://github.com/firebase/firebase-admin-python/pull/531/files#diff-81622636c5d10b6ab40396c608887473086a43b94621195117193ab81a4b5cdaR166 - `emulated` is `False` first, then `True`).

These tests broke due to the changes made to token validation to fit the emulator (https://github.com/firebase/firebase-admin-python/pull/531/commits/aacd27fac95415fc42f74ec744ec0894231d704d) - they were supposed to raise an exception but didn't. Note that all tests ran perfectly fine before that.

hiranya911 commented:
I'm debugging the tests in your branch and I can clearly see that tests that shouldn't be running in the emulator mode, picking up the emulator mode from another fixture. The test case `test_noncert_credential[user_mgt_app0]` should pass by raising the expected error, but it doesn't because it sees the env variable set in a different fixture, and goes into the emulator mode.

muru commented on tests/test_token_gen.py:
```diff
@@ -133,23 +141,39 @@ def _overwrite_iam_request(app, request):
-@pytest.fixture(scope='module')
-def auth_app():
+
+def _is_emulated():
```
The intention is that once the tests are fixed to accomodate the emulator, if the emulator variable name is changed, these tests should break. But I suppose the environment variable name could be a parameter for the function or something.

muru commented on tests/test_token_gen.py:
```diff
@@ -614,6 +665,8 @@ def test_certificate_request_failure(self, user_mgt_app):
 class TestCertificateCaching:
 
     def test_certificate_caching(self, user_mgt_app, httpserver):
+        if _is_emulated():
```
Yes, right now I don't have the understanding of the emulator required to know which of these tests are actually important for the emulator - what behaviour is supported and expected and what behaviour is undefined. The failing tests are all marked skipped for the emulator case here for review (so that the reviewer doesn't have to jump back and forth between test results and code). If somebody can triage them and note the important ones, I can then prioritise fixing those.

hiranya911 replied:
Ideally everything except a few should pass on both modes. Some of the ones that are not relevant in the emulator mode are:

* test_sign_with_iam
* test_sign_with_iam_error
* test_sign_with_discovered_service_account
* test_sign_with_discovery_failure
* test_certificate_request_failure
* test_certificate_caching

muru commented on integration/test_auth.py:
```diff
@@ -32,11 +33,18 @@
-_password_reset_url = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty/resetPassword'
-_verify_email_url = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty/setAccountInfo'
-_email_sign_in_url = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty/emailLinkSignin'
+_EMULATOR_HOST_ENV_VAR = 'FIREBASE_AUTH_EMULATOR_HOST'
```
I suppose the question here is whether the knowledge that some tests do pass worth marking the failing tests as being skipped for emulator (they should still be run for the "real" integration tests). At the least, this does reduce future work a bit since we'll already know which tests to go about fixing, and *might* catch a breakage in the meantime.

hiranya911 replied:
It's a lot of knowledge to capture in one PR. So let's move that into a separate PR, and keep this one focused on the implementation.

We also encountered some flaky tests when we tried this in Node.js, which is why that PR is still on hold. So a passing integration test doesn't necessarily mean everything is ok either.

muru commented:
Ah, I hadn't noticed that, thanks! I'll look into how that can be fixed - maybe have an existing value of the environment variable override the fixture parameters.

hiranya911 commented:
I think the easiest solution would be to change the fixture scope to `function` (which is also pytest default). That will cause a new app and a monkeypatch instance to be created per test function, and they will get cleaned up immediately after. App initialization and clean up are pretty cheap operations, so it should be ok in terms of speed.

hiranya911 commented:
With function-scoped fixtures (plus a few other minor modifications), these are all the remaining test failures:

```
FAILED tests/test_token_gen.py::TestCreateCustomToken::test_noncert_credential[user_mgt_app1] - Failed: DID NOT RAISE <class 'ValueError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_revoked_token_check_revoked[user_mgt_app1-BinaryToken] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.Revoke...
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_revoked_token_check_revoked[user_mgt_app1-TextToken] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.RevokedI...
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-NoKid] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-WrongKid] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-FutureToken] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_invalid_token[user_mgt_app1-ExpiredToken] - Failed: DID NOT RAISE <class 'firebase_admin._auth_utils.InvalidIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_expired_token[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.ExpiredIdTokenError'>
FAILED tests/test_token_gen.py::TestVerifyIdToken::test_certificate_request_failure[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.CertificateFetchEr...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-NoKid] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCooki...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-WrongKid] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessionCo...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-FutureCookie] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSessi...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_invalid_cookie[user_mgt_app1-ExpiredCookie] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.InvalidSess...
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_expired_cookie[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.ExpiredSessionCookieError'>
FAILED tests/test_token_gen.py::TestVerifySessionCookie::test_certificate_request_failure[user_mgt_app1] - Failed: DID NOT RAISE <class 'firebase_admin._token_gen.CertificateF...
FAILED tests/test_token_gen.py::TestCertificateCaching::test_certificate_caching[user_mgt_app1] - assert 0 == 1
```

It's unfortunate our implementation doesn't handle expired emulated tokens correctly (this is because we bypass the JWT verification step entirely for emulated tokens). But this might be ok. I'm fine with skipping the above tests if we must.

muru commented:
@hiranya911 I removed the integration test and modified the fixtures to use the `function` scope, and where possible, I modified the tests to account for the current behaviour of the emulator mode instead of skipping them. If the emulator mode is improved, that might help prevent tests from being overlooked due to being skipped.

Also, I opted against using a module-level variable for `_auth_utils.is_emulated()` as it behaves like module fixture when a variable, causing test failures. Is that fine?

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -286,6 +347,14 @@ def test_sign_with_discovery_failure(self):
     def _verify_signer(self, token, signer):
         segments = token.split('.')
         assert len(segments) == 3
+        # See https://github.com/googleapis/google-auth-library-python/pull/324
```
What if we just did a `jwt.decode(verify=False)` here instead?

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -212,20 +244,30 @@ def test_invalid_params(self, auth_app, values):
 
     def test_sign_with_iam(self):
-        options = {'serviceAccountId': 'test-service-account', 'projectId': 'mock-project-id'}
+        signer = _token_gen.AUTH_EMULATOR_EMAIL if _is_emulated() else 'test-service-account'
```
This test doesn't use any of the fixtures. Does it ever run in the emulator mode?

Update: I see a few other tests in this file that doesn't depend on the fixtures, and therefore never runs in the emulator mode. Let's not modify them.

muru replied:
It might not use those fixtures, but the presence of those fixtures probably affects it via the environment variable. At any rate, I did get failures for this and the tests below, where the email address in the assertion failure definitely was the one used by the emulator code.

hiranya911 replied:
That sounds like a bug to me. With function-scoped fixtures, these tests should never see the emulator mode enabled.

muru replied:
Ah, yes, with the change to function-scoped fixtures, they no longer see emulator mode enabled.

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -440,17 +515,31 @@ def test_invalid_arg(self, user_mgt_app, id_token):
         assert excinfo.value.http_response is None
 
     def test_expired_token(self, user_mgt_app):
+        _overwrite_cert_request(user_mgt_app, MOCK_REQUEST)
```
Redundant?

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -440,17 +515,31 @@ def test_invalid_arg(self, user_mgt_app, id_token):
+    def test_invalid_token(self, user_mgt_app, id_token_key):
+        id_token = self.invalid_tokens[id_token_key]
+        if _is_emulated() and id_token_key in self.tokens_not_invalid_in_emulator:
+            # These tokens won't be invalid when using the emulator, check them like valid tokens.
```
There's quite a bit of duplication and commentary here. Let's try to reduce that as much as possible. For example:

```
if _is_emulated() and id_token_key in self.tokens_not_invalid_in_emulator:
    return _assert_valid_in_emulator(id_token, user_mgt_app)
```

Then you can use the same helper function in places like `test_expired_token`.

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -560,8 +659,15 @@ def test_invalid_args(self, user_mgt_app, cookie):
+    def test_invalid_cookie(self, user_mgt_app, cookie_key):
+        cookie = self.invalid_cookies[cookie_key]
+        if _is_emulated() and cookie_key in self.cookies_not_invalid_in_emulator:
+            # These cookies won't be invalid when using the emulator, check them like valid cookies.
```
Same here. Extract duplicate code into helper functions. Avoid comments.

hiranya911 reviewed (changes requested):
Thanks for making the changes @muru. I think this looks pretty reasonable. I've left some comments on how to further clean up and reduce some of the code duplication.

hiranya911 commented on firebase_admin/_token_gen.py:
```diff
@@ -291,14 +311,16 @@ def verify(self, token, request):
                 '{0} expects {1}, but was given a custom '
                 'token.'.format(self.operation, self.articled_short_name))
         elif not header.get('kid'):
+            if _auth_utils.is_emulated():
```
Use a local variable please. Do not call `is_emulated()` more than once per verification.

hiranya911 commented on firebase_admin/_token_gen.py:
```diff
@@ -285,20 +305,24 @@ def verify(self, token, request):
                 '{0} expects {1}, but was given a custom '
                 'token.'.format(self.operation, self.articled_short_name))
         elif not header.get('kid'):
+            if emulated:
```
```
elif not emulated and not header.get('kid'):
```

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -522,20 +579,27 @@ class TestVerifySessionCookie:
         'IDToken': TEST_ID_TOKEN,
     }
 
+    cookies_not_invalid_in_emulator = [
```
Let's call this `invalid_cookies_accepted_in_emulator` to avoid the double negative.

muru replied:
That's what I originally had, but it breaks the linter as the name's too long:
```
tests/test_token_gen.py:428:4: C0103: Class attribute name "invalid_tokens_accepted_in_emulator" doesn't conform to '([A-Za-z_][A-Za-z0-9_]{2,30}|(__.*__))$' pattern (invalid-name)
tests/test_token_gen.py:583:4: C0103: Class attribute name "invalid_cookies_accepted_in_emulator" doesn't conform to '([A-Za-z_][A-Za-z0-9_]{2,30}|(__.*__))$' pattern (invalid-name)
```

(Same problem with the slightly shorter variant `invalid_tokens_valid_in_emulator`)

hiranya911 replied:
How about just `cookies_accepted_in_emulator` and `tokens_accepted_in_emulator`?

muru replied:
Still feels not quite right to me, but I don't want to bikeshed on this, so `*_accepted_in_emulator` it is.

hiranya911 replied:
I understand the concern. It's a compromise for sure.

hiranya911 commented on tests/test_token_gen.py:
```diff
@@ -388,14 +425,23 @@ class TestVerifyIdToken:
-    def test_valid_token(self, user_mgt_app, id_token):
-        _overwrite_cert_request(user_mgt_app, MOCK_REQUEST)
-        claims = auth.verify_id_token(id_token, app=user_mgt_app)
+    tokens_not_invalid_in_emulator = [
```
invalid_tokens_accepted_in_emulator

hiranya911 commented on tests/test_user_mgt.py:
```diff
@@ -50,17 +50,30 @@
 
-@pytest.fixture(scope='module')
-def user_mgt_app():
+@pytest.fixture(scope='module', params=[{'emulated': False}, {'emulated': True}])
```
To be on the safe side should we change this to function scope as well?

muru replied:
Done. Test times went from 7-7.5s to 7.5-8s. Not much of a problem, I think.

hiranya911 reviewed (changes requested):
Thanks @muru. This looks great! I only had a few minor comments, and we can merge once those are addressed.

hiranya911 reviewed (approved):
Thanks @muru for patiently working through all the comments. This LGTM 👍
